hudi-client/hudi-client-common/src/main/java/org/apache/hudi/optimize/HilbertCurve.java

```diff
@@ -0,0 +1,321 @@
+    }
+    if (index.signum() == -1) {
+      throw new IllegalArgumentException("index cannot be negative");
+    }
+    return transposedIndexToPoint(bits, transpose(index));
+  }
+
+  public void point(BigInteger index, long[] x) {
+    if (index == null) {
+      throw new NullPointerException("index must not be null");
+    }
+    if (index.signum() == -1) {
+      throw new IllegalArgumentException("index cannot be negative");
+    }
+    Arrays.fill(x, 0);
+    transpose(index, x);
+    transposedIndexToPoint(bits, x);
+  }
+
+  public void point(long i, long[] x) {
+    point(BigInteger.valueOf(i), x);
+  }
+
+  /**
+   * Converts a {@code long} index (distance along the Hilbert Curve from 0) to a
+   * point of dimensions defined in the constructor of {@code this}.
+   *
+   * @param index
+   *            index along the Hilbert Curve from 0. Maximum value 2
+   *            <sup>bits+1</sup>-1.
+   * @return array of longs being the point
+   * @throws IllegalArgumentException
+   *             if index is negative
+   */
+  public long[] point(long index) {
+    return point(BigInteger.valueOf(index));
+  }
+
+  /**
+   * Returns the transposed representation of the Hilbert curve index.
+   *
+   * <p>
+   * The Hilbert index is expressed internally as an array of transposed bits.
+   *
+   * <pre>
+   Example: 5 bits for each of n=3 coordinates.
+   15-bit Hilbert integer = A B C D E F G H I J K L M N O is stored
+   as its Transpose                        ^
+   X[0] = A D G J M                    X[2]|  7
+   X[1] = B E H K N        &lt;-------&gt;       | /X[1]
+   X[2] = C F I L O                   axes |/
+   high low                         0------&gt; X[0]
+   * </pre>
+   *
+   * @param index
+   *            index to be tranposed
+   * @return transposed index
+   */
+  long[] transpose(BigInteger index) {
+    long[] x = new long[dimensions];
+    transpose(index, x);
+    return x;
+  }
+
+  private void transpose(BigInteger index, long[] x) {
+    byte[] b = index.toByteArray();
+    for (int idx = 0; idx < 8 * b.length; idx++) {
+      if ((b[b.length - 1 - idx / 8] & (1L << (idx % 8))) != 0) {
+        int dim = (length - idx - 1) % dimensions;
+        int shift = (idx / dimensions) % bits;
+        x[dim] |= 1L << shift;
+      }
+    }
+  }
+
+  /**
+   * <p>
+   * Given the axes (coordinates) of a point in N-Dimensional space, find the
+   * distance to that point along the Hilbert curve. That distance will be
+   * transposed; broken into pieces and distributed into an array.
+   *
+   * <p>
+   * The number of dimensions is the length of the hilbertAxes array.
+   *
+   * <p>
+   * Note: In Skilling's paper, this function is called AxestoTranspose.
+   *
+   * @param bits
+   * @param point
+   *            Point in N-space
+   * @return The Hilbert distance (or index) as a transposed Hilbert index
+   */
+  static long[] transposedIndex(int bits, long... point) {
```

Comment:
Let's make sure we have tests for these utility methods

Reply:
ok, will add UT.